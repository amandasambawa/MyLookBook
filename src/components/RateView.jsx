import React, { Component } from 'react';
import { database } from '../firebase.js';
import { Redirect } from 'react-router-dom';
import Rate from 'rc-rate';
import "../styles/stars.css";
import AlertContainer from 'react-alert';
import "../styles/RateView.css"

class RateView extends Component {

  constructor(props){
    super(props);
    this.state = {
      ratingComposition: 0,
      ratingTrendy: 0,
      ratingComment: "",
      haveSaved: false
    }
      this.handleComposition = this.handleComposition.bind(this);
      this.handleTrendy = this.handleTrendy.bind(this);
      this.saveRating = this.saveRating.bind(this);
      this.handleUsernameChange = this.handleUsernameChange.bind(this);
  }


  componentDidMount() {
    //grab outfit image in database
    let image = null;
    database.ref(`/users/${this.props.match.params.userId}/outfitobjects/${this.props.match.params.outfitId}`)
    .once("value").then((snapshot)=> {
        image = snapshot.child("img").val();
        this.setState({ outfitImage: image });
    });
  }


  //the alert options for the npm react-alert
  alertOptions = {
    offset: 14,
    position: 'top right',
    theme: 'dark',
    time: 5000,
    transition: 'scale'
  }

  //handles the composition rating
  handleComposition(rating) {
    if(this.state.haveSaved === false){
      this.setState({
        ratingComposition: rating
      })
    }
  }

  //handles the trendy rating
  handleTrendy(rating) {
    if(this.state.haveSaved === false){
      this.setState({
        ratingTrendy: rating
      })
    }
  }

  //handles the username state.
  handleUsernameChange(event){
      this.setState({ratingComment: event.target.value});
  }
  //handles saving the rating and comments
  saveRating() {
    //if the user has not saved once before, we will send that data to the db
    console.log(this.props.match.params);
    if(this.state.haveSaved === false){
        let ratingRef = database.ref(`/users/${this.props.match.params.userId}/outfitobjects/${this.props.match.params.outfitId}/ratings/`);
        let newRating = ratingRef.push();

        newRating.set({
          comment: this.state.ratingComment,
          composition: this.state.ratingComposition,
          trendy: this.state.ratingTrendy
        })

        this.setState({haveSaved: true});
    } else {
      //the user has already pressed save, give them alert denying the option
      this.msg.show('You have already rated this image!', {
        offset: 14,
        position: 'top right',
        theme: 'dark',
        time: 5000,
        transition: 'scale',
        type: null
      })

    }
  }


  render(){
    if (this.props.uid){
      return <Redirect to={{ pathname: `/singleOutfit/${this.props.match.params.outfitId}` }} />
    }else{
      return(
        <div id="rateViewContainer">

          <div className ="imageIDContainer">
            <img className="imageID" src={this.state.outfitImage}/>
          </div>

          <div className="ratingsLabel">Composition</div>
          <Rate
             defaultValue={0}
             onChange={this.handleComposition}
             style={{ fontSize: 40 }}
           />

          <div className="ratingsLabel">Trendy</div>
          <Rate
              defaultValue={0}
              onChange={this.handleTrendy}
              style={{ fontSize: 40 }}
            />

          <div className="ratingsLabel">Comment</div>
          <textarea placeholder="Leave a comment!" onChange={this.handleUsernameChange}></textarea>

          <AlertContainer ref={a => this.msg = a} {...this.alertOptions} />
          <button onClick={this.saveRating} className="button">Save</button>

        </div>
      );
    }

  }

}

export default RateView;
