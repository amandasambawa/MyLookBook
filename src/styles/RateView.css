#rateViewContainer {
    text-align: center;
}

.imageID{
  height: 300px;
  width: 90%;
  text-align: center;
}

.imageIDContainer {
    text-align: center;
}

.ratingsLabel {
    font-size: 20px;
    width: 90%;
    margin: auto;
    margin-top: 5px;
}

textarea {
    max-width: 90%;
    margin: auto;
    margin-bottom: 10px;
}