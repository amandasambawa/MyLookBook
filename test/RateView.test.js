import React from 'react';
import ReactDOM from 'react-dom';
import {shallow, render, mount} from 'enzyme';
import RateView from '../src/components/RateView.jsx';
import Rate from 'rc-rate';


/*
* This test file is for RateView
*
*/

//Make sure that RateView renders properly.
it('RateView renders properly.', () => {
  const wrapper = shallow(<RateView testing={true}/>);
});



/*
* The following tests encompass the UI.
*/


//Make sure the Composition UI is loading properly
it('renders the Comment Section of the page', () => {
  const wrapper = shallow(<RateView testing={true}/>);
  const title =           <div className="ratingsLabel">
            Composition
          </div>;
  expect(wrapper.contains(title)).toEqual(true);
});

//Make sure the Trendy UI is loading properly.
it('renders the Trendy section of the page', () => {
  const wrapper = shallow(<RateView testing={true}/>);
  const title =           <div className="ratingsLabel">
            Trendy
          </div>;
  expect(wrapper.contains(title)).toEqual(true);
});


//Make sure the Comment UI is loading properly
it('renders the Comment Section of the page', () => {
  const wrapper = shallow(<RateView testing={true}/>);
  const title =           <div className="ratingsLabel">
            Comment
          </div>;
  expect(wrapper.contains(title)).toEqual(true);
});



//make sure the images load space in the webpage.
describe('RateView', () => {
    it('check if the images loading up', () => {
        //let obj = {params: {outfitId: `-KkI2QGzORcXRcSoHT9j`} };
        const wrapper = shallow(<RateView testing={true}/>);
        const image = wrapper.find('img');
        expect(image.prop('src')).toEqual(true);

    });
  });


/*
* The following tests encompass the functionality
*/

  //Make sure handleComposition is changing the handleComposition state properly.
  it('make sure that handleComposition is working properly.', () => {
    const wrapper = shallow(<RateView testing={true}/>);
    wrapper.instance().handleComposition(5);
    expect(wrapper.state('ratingComposition')).toEqual(5);
  });

  //Make sure handleTrendy is changing the handleComposition state properly.
  it('make sure that handleTrendy is working properly.', () => {
    const wrapper = shallow(<RateView testing={true}/>);
    wrapper.instance().handleTrendy(1);
    expect(wrapper.state('ratingTrendy')).toEqual(1);
  });

  //Make sure handleCommentChange is changing the ratingComment state properly.
  it('make sure that handleCommentChange is working properly.', () => {
    const wrapper = mount(<RateView testing={true}/>);
    wrapper.instance().setState({ratingComposition:'hello world'});
    expect(wrapper.state('ratingComposition')).toEqual('hello world');
  });
