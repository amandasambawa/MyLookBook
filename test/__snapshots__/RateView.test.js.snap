exports[`test renders correctly 1`] = `
<div>
  <h1>
    RateView
  </h1>
  <div
    className="imageIDContainer">
    <img
      className="imageID"
      src={undefined} />
  </div>
  <h2>
    Composition
  </h2>
  <ul
    className="rc-rate"
    onMouseLeave={[Function]}
    style={
      Object {
        "fontSize": 40,
      }
    }>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
  </ul>
  <h2>
    Trendy
  </h2>
  <ul
    className="rc-rate"
    onMouseLeave={[Function]}
    style={
      Object {
        "fontSize": 40,
      }
    }>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
    <li
      className="rc-rate-star rc-rate-star-zero"
      onClick={[Function]}
      onMouseMove={[Function]}>
      <div
        className="rc-rate-star-first">
        ★
      </div>
      <div
        className="rc-rate-star-second">
        ★
      </div>
    </li>
  </ul>
  <h2>
    Comment
  </h2>
  <textarea
    onChange={[Function]}
    placeholder="Leave a comment!" />
  <div
    className="css-132hx50">
    <span />
  </div>
  <button
    className="button"
    onClick={[Function]}>
    Save
  </button>
</div>
`;
